feat: action bar on selected blocks

tiensonqin opened the pull request:
This PR introduces a new action bar for selected blocks, it also works on table view.

<img width="645" alt="image" src="https://github.com/user-attachments/assets/507d287e-fc5e-4642-bd26-864e93a97878" />

<img width="661" alt="image" src="https://github.com/user-attachments/assets/0914ff60-5bb0-4615-a8ca-8d86877c0010" />

RCmerci reviewed (commented):
batch set property not working properly.
1. select 2 blocks
2. click 'set property', select a 'icon'
3. only the first block has the icon

RCmerci reviewed (commented):
After selecting multiple blocks using 'shift+<mouse-click>', the action bar does not appear

logseq-cldwalker commented:
@tiensonqin A couple of bugs I'm seeing:
* Like Zhiyuan I'm not seeing an action bar when selecting blocks with mouse or keyboard
* Changing the view type from a table no longer works e.g. table from a property page
* Setting properties from a table doesn't allow selecting a property value. It seems to only create an empty property value. This feels buggy especially since the table doesn't update to show the changes or there's not at least a notification saying anything changed
* Cutting from a property table isn't undoable. Should we rename cutting to delete? I was surprised this action  deleted properties from the selected rows
* When doing a batch action from table, I can create a new property but I can't create a new tag. Is that intentional?

Let me know when I should take another look. Also, is there a way to unset a property for selected rows or does that come later?

tiensonqin commented:
@cldwalker 
> Like Zhiyuan I'm not seeing an action bar when selecting blocks with mouse or keyboard

It's intentional not to show an action bar with keyboard shortcuts (mod+up/down) because it's distracting and showing an action bar will prevent any further actions in our current implementation (using radix UI).
I think you need to run `yarn` in packages/ui to see the action bar when selecting blocks with mouse, please record a video if it doesn't work.

> Changing the view type from a table no longer works e.g. table from a property page
> Setting properties from a table doesn't allow selecting a property value. It seems to only create an empty property value. This feels buggy especially since the table doesn't update to show the changes or there's not at least a notification saying anything changed

I can't reproduce those two bugs, can you provide some reproduce steps? Thanks.

> Cutting from a property table isn't undoable. Should we rename cutting to delete? I was surprised this action deleted properties from the selected rows

Cutting from any table will delete the blocks, we can tweak the behavior to remove the current property from the selected blocks if you think that's expected.

> When doing a batch action from table, I can create a new property but I can't create a new tag. Is that intentional?

No, it's a bug, I'll push a fix soon.

tiensonqin commented:
@logseq-cldwalker Thanks for the videos and reproducing steps!

> That make sense not to show automatically. It would be helpful for us keyboard users if the actions bar could somehow be activated after selecting blocks e.g. from the right click menu

I updated the behavior to show action bar after selecting blocks with keyboard (shift+up/down), it'll be hidden if there's any further action such as selecting more blocks or using arrows for navigation. I didn't enable action bar after pressing `ESC` to select a block because I still find it distracting. 
I also want to set a shortcut to focus on the action bar when there're selected blocks, any ideas? cc @RCmerci @xyhp915 

> When I try to use it on two blocks, none of the property types batch add except for :checkbox -

Fixed.

> Changing the view type from a table no longer works e.g. table from a property page

I still can't reproduce this one.

> Setting properties from a table doesn't allow selecting a property value. It seems to only create an empty property value.

Fixed.

> The ux for batch adding from the table feels buggier compared to selecting from blocks. After a batch action in the table, the rows are still selected which makes it feel like the action never took place. Also, none of the added properties display in a table, even if I navigate away and back

Agreed. The added properties are not displayed because only predefined properties are displayed for tags and properties (al properties are shown for query results), showing all properties requires loading all the rows, which may affect the view performance a lot. I plan to revisit this after #11774.

> I found the delete button to be better ux b/c I understood I was deleting entities. With cut, I expect undo to work like it does in a normal block view i.e. undo and the cutted blocks reappear in the table. Currently we have to navigate away and back to see the undone table updated

I updated `cut` to a `trash` button. There're some general issues with table reactivity, including undo/delete/insert, again, I'd like to address those issues after #11774 because view data loading will be changed a lot in #11774.

tiensonqin commented:
> Also, is there a way to unset a property for selected rows or does that come later?

You can unset :default, :url, :number, :node
<img width="555" alt="image" src="https://github.com/user-attachments/assets/c70a5ca4-a40b-4afb-a867-840673ad2612" />

And :data, :datetime
<img width="453" alt="image" src="https://github.com/user-attachments/assets/6f64e941-2ac8-4d8f-a279-87b41b752709" />

I'm still thinking of a universal approach to handle all types including :checkbox. Let me know if anyone has some ideas.

tiensonqin commented:
@RCmerci The above two bugs are fixed.

tiensonqin commented:
> Changing the view type from a table no longer works e.g. table from a property page

This one should be fixed by https://github.com/logseq/logseq/pull/11763/commits/170edfb772876268789c41d84005c0a2b3fc2110.

logseq-cldwalker commented:
> I updated the behavior to show action bar after selecting blocks with keyboard (shift+up/down), it'll be hidden if there's any further action such as selecting more blocks or using arrows for navigation. I didn't enable action bar after pressing ESC to select a block because I still find it distracting.

Thanks!

> > When I try to use it on two blocks, none of the property types batch add except for :checkbox -
> Fixed.

Thanks for fixing most of these. I'm still seeing :default :many and :url :many as not able to batch set - https://www.loom.com/share/db31921682df4049a0cce97aa1cfadd6. Fwiw I just use the properties example graph to test all these property combinations

> Agreed. The added properties are not displayed because only predefined properties are displayed for tags and properties (al properties are shown for query results), showing all properties requires loading all the rows, which may affect the view performance a lot. I plan to revisit this after https://github.com/logseq/logseq/pull/11774.

Good to know we're still thinking about this. If it's not happening soon, it would be helpful to just display a notification e.g. `Properties added!` as the current ux does still feel unresponsive. We don't have this problem with tags because they are always visible

> I updated cut to a trash button. There're some general issues with table reactivity, including undo/delete/insert, again, I'd like to address those issues after https://github.com/logseq/logseq/pull/11774 because view data loading will be changed a lot in https://github.com/logseq/logseq/pull/11774.

Good to know. Thanks for the change for now

> This one should be fixed by https://github.com/logseq/logseq/commit/170edfb772876268789c41d84005c0a2b3fc2110.

I haven't seen this so I think it's fixed

I'm not sure if this is a bug: when I copy on a batch of nodes that include pages, only the blocks get copied - https://www.loom.com/share/941d8989061646908b5dea6a054418ed . If copy can't handle pages maybe it would be helpful to have export edn at some point since it can handle any block type

Just this question and the bugs on :url :many and :default :many left

logseq-cldwalker commented:
One other bug I noticed, the property values batch added with a  :default or :url property are the same entity, causing the property values to be in sync when edited:

<img width="366" alt="Screen Shot 2025-03-11 at 12 37 21 PM" src="https://github.com/user-attachments/assets/e5fd6730-81b3-4649-af1f-2c236ef32ba8" />

tiensonqin commented:
@logseq-cldwalker I've fixed all the issues mentioned above.

>  when I copy on a batch of nodes that include pages, only the blocks get copied -

Page titles will be copied.

tiensonqin commented:
> With multiple blocks selected, batch add :checkbox properties. Only the first block has the property added

Can you record a video? I'm not able to reproduce this one.

> The 2nd bug is tricky because the :many dropdown almost seems in conflict with the data model

Indeed, there are also some cases where it makes sense to set the same block as the property value for multiple blocks. Let's resolve those edge cases when we have more feedback from users.

>  I'll be sure to mention removing properties is found in property values in the docs. 

I added `Unset property`.

logseq-cldwalker commented:
> Can you record a video? I'm not able to reproduce this one.

Confirmed :checkbox is fixed now

> Indeed, there are also some cases where it makes sense to set the same block as the property value for multiple blocks. Let's resolve those edge cases when we have more feedback from users.

Sounds good. I'll create a bug for https://www.loom.com/share/66cfd0e23104418eae80fa96139325c0 

> I added Unset property.

Thanks!

logseq-cldwalker reviewed (approved):
@tiensonqin Exciting work! 👍 🚀 Thanks for handling so many permutations of property types. Fixed two small bugs I noticed at the end. One last bug I saw is that batch setting icons doesn't work. I didn't see any other public built-in properties have issues
